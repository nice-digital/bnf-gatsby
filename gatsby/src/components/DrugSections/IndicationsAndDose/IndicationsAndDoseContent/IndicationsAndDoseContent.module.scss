@use 'sass:color';
@use '../../../../styles/colours';
@use '@nice-digital/nds-core/scss/spacing';
@use '@nice-digital/nds-core/scss/typography';
@use '@nice-digital/nds-core/scss/utils';

.indicationWrapper {
  background: colours.$indication-content-wrapper-background;
  border: 1px solid colours.$indication-content-wrapper-border;
  margin-bottom: utils.rem(spacing.$medium);
  padding: utils.rem(spacing.$medium);
}

.indicationHeading {
  break-inside: avoid;
  line-height: 1.3;
  margin: utils.rem(0 0 spacing.$medium);
  max-width: none;
}

.indicationText {
  display: block;
  margin-top: utils.rem(spacing.$small);

  &:first-child {
    margin-top: 0;
  }
}

.routeOfAdministration {
  margin-bottom: utils.rem(spacing.$medium);

  &:last-child {
    margin-bottom: 0;
  }
}

.routeOfAdministrationHeading {
  @include typography.p;
  break-inside: avoid;
  margin: utils.rem(0 0 spacing.$medium);
  max-width: none;
}

.patientGroups {
  margin: 0;
}

.patientGroupDose {
  break-inside: avoid;
  margin-bottom: utils.rem(spacing.$medium);
  padding-left: utils.rem(spacing.$medium);
  position: relative;

  &:last-child {
    margin-bottom: 0;
  }

  &::before {
    content: '';
    height: 100%;
    left: 0;
    position: absolute;
    top: 0;
    width: utils.rem(spacing.$small);
  }

  /* stylelint-disable function-parentheses-space-inside */
  // Use patterned colours to distinguish patient groups
  &.neonate::before {
    background: repeating-linear-gradient(
      -45deg,
      colours.$indication-dose-neonate-1,
      colours.$indication-dose-neonate-1 1px,
      colours.$indication-dose-neonate-2 1px,
      colours.$indication-dose-neonate-2 12px
    );
  }

  &.child::before {
    background: repeating-linear-gradient(
      45deg,
      color.adjust(colours.$indication-dose-child, $lightness: 75%),
      color.adjust(colours.$indication-dose-child, $lightness: 75%) 1px,
      colours.$indication-dose-child 1px,
      colours.$indication-dose-child 12px
    );
  }

  &.adult::before {
    background: colours.$indication-dose-adult;
  }
}

.patientGroup {
  font-weight: bold;
}

.doseStatement {
  margin: 0;
}

.supplementary {
  > :last-child {
    margin-bottom: 0;
  }
}
