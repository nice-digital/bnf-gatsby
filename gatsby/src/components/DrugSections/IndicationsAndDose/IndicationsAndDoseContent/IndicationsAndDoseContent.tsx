import slugify from "@sindresorhus/slugify";
import classNames from "classnames";
import { Fragment, type FC, useMemo } from "react";
import striptags from "striptags";

import { type FeedIndicationsAndDosePotContent } from "@nice-digital/gatsby-source-bnf";

import { Accordion } from "@/components/Accordion/Accordion";
import { QueryResult } from "@/utils";

import styles from "./IndicationsAndDoseContent.module.scss";

export interface IndicationsAndDoseContentProps {
	content: QueryResult<FeedIndicationsAndDosePotContent>;
	/** The string to be rendered before the `contentFor` field.
	 *
	 * Defaults to `For`. Pass `For all` for drug classes to read `For all tetracyclines` for example. */
	contentForPrefix?: "For" | "For all";
	/** Whether to (`true`) or not (`false`) to wrap this indications and dose content in an accordion to make it collapsible */
	collapsible: boolean;
	/** Whether to open the accordion (if `collapsible=true`) by default */
	defaultOpen: boolean;
}

export const IndicationsAndDoseContent: FC<IndicationsAndDoseContentProps> = ({
	content: {
		contentFor,
		indicationAndDoseGroups,
		doseEquivalence,
		doseAdjustments,
		extremesOfBodyWeight,
		pharmacokinetics,
		potency,
	},
	contentForPrefix = "For",
	collapsible,
	defaultOpen,
}) => {
	const slug = slugify(striptags(contentFor));

	const body = useMemo(
		() => (
			<>
				{indicationAndDoseGroups?.map(
					({ therapeuticIndications, routesAndPatientGroups }, groupIndex) => {
						const groupId = `${slug}-indication-${groupIndex + 1}`;

						return (
							<section
								key={groupIndex}
								className={styles.indicationWrapper}
								aria-labelledby={groupId}
							>
								<h4 id={groupId} className={styles.indicationHeading}>
									{therapeuticIndications.map(
										(
											{ indication, sctIndication, sctTherapeuticIntent },
											indicationIndex
										) => (
											<Fragment key={indication}>
												<span
													dangerouslySetInnerHTML={{
														__html:
															indication +
															(indicationIndex <
															therapeuticIndications.length - 1
																? ", "
																: ""),
													}}
													className={styles.indicationText}
													data-sct-indication={sctIndication}
													data-sct-therapeutic-intent={sctTherapeuticIntent}
												/>
											</Fragment>
										)
									)}
									<span
										className="visually-hidden"
										dangerouslySetInnerHTML={{ __html: ` for ${contentFor}` }}
									/>
								</h4>
								{routesAndPatientGroups.map(
									({ routeOfAdministration, patientGroups }) => (
										<section
											key={routeOfAdministration}
											className={styles.routeOfAdministration}
										>
											<h5 className={styles.routeOfAdministrationHeading}>
												{routeOfAdministration}
											</h5>
											<dl className={styles.patientGroups}>
												{patientGroups.map(
													({
														patientGroup,
														detailedPatientGroup,
														doseStatement,
													}) => (
														<div
															key={patientGroup}
															className={classNames(
																styles.patientGroupDose,
																styles[patientGroup]
															)}
														>
															<dt className={styles.patientGroup}>
																{detailedPatientGroup}
															</dt>
															<dd
																className={styles.doseStatement}
																dangerouslySetInnerHTML={{
																	__html: doseStatement,
																}}
															/>
														</div>
													)
												)}
											</dl>
										</section>
									)
								)}
							</section>
						);
					}
				)}
				{doseAdjustments && (
					<section aria-labelledby={`${slug}-dose-adjustments`}>
						<h4 id={`${slug}-dose-adjustments`}>
							Dose adjustments due to interactions
							<span
								className="visually-hidden"
								dangerouslySetInnerHTML={{ __html: ` for ${contentFor}` }}
							/>
						</h4>
						<div
							className={styles.supplementary}
							dangerouslySetInnerHTML={{ __html: doseAdjustments }}
						/>
					</section>
				)}
				{doseEquivalence && (
					<section aria-labelledby={`${slug}-dose-equivalence`}>
						<h4 id={`${slug}-dose-equivalence`}>
							Dose equivalence and conversion
							<span
								className="visually-hidden"
								dangerouslySetInnerHTML={{ __html: ` for ${contentFor}` }}
							/>
						</h4>
						<div
							className={styles.supplementary}
							dangerouslySetInnerHTML={{ __html: doseEquivalence }}
						/>
					</section>
				)}
				{extremesOfBodyWeight && (
					<section aria-labelledby={`${slug}-extremes-of-body-weight`}>
						<h4 id={`${slug}-extremes-of-body-weight`}>
							Doses at extremes of body-weight
							<span
								className="visually-hidden"
								dangerouslySetInnerHTML={{ __html: ` for ${contentFor}` }}
							/>
						</h4>
						<div
							className={styles.supplementary}
							dangerouslySetInnerHTML={{ __html: extremesOfBodyWeight }}
						/>
					</section>
				)}
				{potency && (
					<section aria-labelledby={`${slug}-potency`}>
						<h4 id={`${slug}-potency`}>
							Potency
							<span
								className="visually-hidden"
								dangerouslySetInnerHTML={{ __html: ` for ${contentFor}` }}
							/>
						</h4>
						<div
							className={styles.supplementary}
							dangerouslySetInnerHTML={{ __html: potency }}
						/>
					</section>
				)}
				{pharmacokinetics && (
					<section aria-labelledby={`${slug}-pharmacokinetics`}>
						<h4 id={`${slug}-pharmacokinetics`}>
							Pharmacokinetics
							<span
								className="visually-hidden"
								dangerouslySetInnerHTML={{ __html: ` for ${contentFor}` }}
							/>
						</h4>
						<div
							className={styles.supplementary}
							dangerouslySetInnerHTML={{ __html: pharmacokinetics }}
						/>
					</section>
				)}
			</>
		),
		[
			indicationAndDoseGroups,
			slug,
			contentFor,
			doseAdjustments,
			doseEquivalence,
			extremesOfBodyWeight,
			potency,
			pharmacokinetics,
		]
	);

	return (
		<section aria-labelledby={`${slug}-indications-and-dose`}>
			{collapsible ? (
				<Accordion
					title={
						<h3
							id={`${slug}-indications-and-dose`}
							dangerouslySetInnerHTML={{
								__html: `${contentForPrefix} ${contentFor}`,
							}}
						/>
					}
					defaultOpen={defaultOpen}
				>
					{body}
				</Accordion>
			) : (
				<>
					<h3
						id={`${slug}-indications-and-dose`}
						className="visually-hidden"
						dangerouslySetInnerHTML={{
							__html: `${contentForPrefix} ${contentFor}`,
						}}
					/>
					{body}
				</>
			)}
		</section>
	);
};
