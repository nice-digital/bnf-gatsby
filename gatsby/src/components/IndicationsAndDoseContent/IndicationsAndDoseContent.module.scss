@use '@nice-digital/nds-core/scss/core';

.indicationWrapper {
  background: core.$nds-colour-nice-white;
  border: 1px solid core.$nds-colour-nice-grey-2;
  margin-bottom: core.rem(core.$nds-spacing-medium);
  padding: core.rem(core.$nds-spacing-medium);

  > :first-child {
    margin-top: 0;
  }

  > :last-child,
  > :last-child > :last-child {
    margin-bottom: 0;
  }
}

.indication {
  display: block;
}

.routeOfAdministration {
  @include core.nds-p;
}

.patientGroupDose {
  border-left: core.rem(core.$nds-spacing-small) solid
    core.$nds-colour-nice-grey-2;
  margin-bottom: core.rem(core.$nds-spacing-medium);
  padding-left: core.rem(core.$nds-spacing-small);

  &:last-child {
    margin-bottom: 0;
  }

  &.neonate {
    border-color: core.$nds-colour-nice-ice-teal;
  }

  &.child {
    border-color: core.$nds-colour-nice-purple;
  }

  &.adult {
    border-color: core.$nds-colour-nice-teal-2;
  }
}

.patientGroup {
  font-weight: bold;
}

.doseStatement {
  margin: 0;
}
