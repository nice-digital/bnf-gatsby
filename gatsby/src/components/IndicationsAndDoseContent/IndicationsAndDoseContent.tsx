import classNames from "classnames";
import { Fragment, type FC } from "react";

import { type FeedIndicationsAndDosePotContent } from "@nice-digital/gatsby-source-bnf";

import styles from "./IndicationsAndDoseContent.module.scss";

export interface IndicationsAndDoseContentProps {
	content: FeedIndicationsAndDosePotContent;
}

export const IndicationsAndDoseContent: FC<IndicationsAndDoseContentProps> = ({
	content: { contentFor, indicationAndDoseGroups },
}) => (
	<section>
		<h3>
			For <span className={styles.contentFor}>{contentFor}</span>
		</h3>
		{indicationAndDoseGroups?.map(
			({ therapeuticIndications, routesAndPatientGroups }, i) => (
				<section key={i} className={styles.indicationWrapper}>
					<h4>
						{therapeuticIndications.map(
							({ indication, sctIndication, sctTherapeuticIntent }, i) => (
								<Fragment key={indication}>
									<span
										dangerouslySetInnerHTML={{ __html: indication }}
										className={styles.indication}
										data-sct-indication={sctIndication}
										data-sct-therapeutic-intent={sctTherapeuticIntent}
									/>
									{i < therapeuticIndications.length - 1 ? (
										<span className={styles.indicationSeparator}> or </span>
									) : null}
								</Fragment>
							)
						)}
					</h4>
					{routesAndPatientGroups?.map(
						({ routeOfAdministration, patientGroups }) => (
							<section key={routeOfAdministration}>
								<h5 className={styles.routeOfAdministration}>
									{routeOfAdministration}
								</h5>
								<dl>
									{patientGroups.map(
										({ patientGroup, detailedPatientGroup, doseStatement }) => (
											<div
												key={patientGroup}
												className={classNames(
													styles.patientGroupDose,
													styles[patientGroup]
												)}
											>
												<dt className={styles.patientGroup}>
													{detailedPatientGroup}
												</dt>
												<dd
													className={styles.doseStatement}
													dangerouslySetInnerHTML={{ __html: doseStatement }}
												/>
											</div>
										)
									)}
								</dl>
							</section>
						)
					)}
				</section>
			)
		)}
	</section>
);
