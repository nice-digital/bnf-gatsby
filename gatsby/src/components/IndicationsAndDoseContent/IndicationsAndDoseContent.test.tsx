import { IndicationsAndDoseContent } from "./IndicationsAndDoseContent";

describe("IndicationsAndDoseContent", () => {
	describe("collapsible", () => {
		it("should wrap content in accordion", () => {});

		it("should render accordion as closed by default", () => {});

		it("should rendering heading 3 with content for and prefix as accordion title", () => {});
	});

	describe("not collapsible", () => {
		it.todo("should not render accordion");

		it.todo("should render visually hidden with content for and prefix");
	});

	describe("body", () => {
		describe("Indications section", () => {
			it.todo("should render section for each indication and dose group");

			it.todo("should label section with indications");

			it.todo("should render heading 4 with multiple indications");
			it.todo(
				"should render SNOMED CT codes as data attributes on indications"
			);
		});

		describe("Routes of administration", () => {
			it.todo("should render section for each route of administration");

			it.todo("should render heading 5 for each route of administration");
		});

		describe("Patient groups", () => {
			it.todo("should render description list for patient groups");
			it.todo("should render item for each patient group");
			it.todo("should render class name for patient group item");
			it.todo("should render patient group with HTML dose statement");
		});
	});
});
