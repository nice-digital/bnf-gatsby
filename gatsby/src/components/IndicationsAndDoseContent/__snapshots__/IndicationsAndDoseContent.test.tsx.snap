// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndicationsAndDoseContent should match snapshot 1`] = `
<section
  aria-labelledby="diazepam-indications-and-dose"
>
  <h3
    class="visually-hidden"
    id="diazepam-indications-and-dose"
  >
    For
     
    diazepam
  </h3>
  <section
    aria-labelledby="diazepam-indication-1"
    class="indicationWrapper"
  >
    <h4
      id="diazepam-indication-1"
    >
      <span
        class="indication"
        data-sct-indication="197480006"
        data-sct-therapeutic-intent="262202000"
      >
        Severe acute anxiety, 
      </span>
      <span
        class="indication"
      >
        Control of acute panic attacks
      </span>
      <span
        class="visually-hidden"
      >
         for 
        diazepam
      </span>
    </h4>
    <section>
      <h5
        class="routeOfAdministration"
      >
        By mouth
      </h5>
      <dl>
        <div
          class="patientGroupDose child"
        >
          <dt
            class="patientGroup"
          >
            Child 1–11 months
          </dt>
          <dd
            class="doseStatement"
          >
            Initially 250 micrograms/kg twice daily.
          </dd>
        </div>
        <div
          class="patientGroupDose child"
        >
          <dt
            class="patientGroup"
          >
            Child 1–4 years
          </dt>
          <dd
            class="doseStatement"
          >
            Initially 2.5 mg twice daily.
          </dd>
        </div>
      </dl>
    </section>
    <section>
      <h5
        class="routeOfAdministration"
      >
        By intramuscular injection, or by slow intravenous injection
      </h5>
      <dl>
        <div
          class="patientGroupDose adult"
        >
          <dt
            class="patientGroup"
          >
            Adult
          </dt>
          <dd
            class="doseStatement"
          >
            10 mg, then 10 mg after at least 4 hours if required, intravenous injection to be administered into a large vein, at a rate of 
            <em>
              not
            </em>
             more than 5 mg/minute.
          </dd>
        </div>
      </dl>
    </section>
  </section>
</section>
`;
